File: AlgosRef/src/Q001_Q049/Q005.java
package Q001_Q049;

public class Q005 {
    public String longestPalindrome(String s) {  // n^2
        if(s==null || s.length()==0) {
            return "";
        }
        int start = 0;
        int end = 0;
        for(int i=0; i<s.length(); i++) {
           int len1 = expand(s, i, i);  // expand to left and right
           int len2 = expand(s, i, i+1);
           int len = Math.max(len1, len2);
           if(len>end-start+1) {  // this is just normal len comparison (+1 is ok)
               start = i-(len-1)/2;  // from current i to left (half len), len-1 is just an adjustment
               end = i+(len/2); // from cuurent i to right (half len)
           }
        }
        return s.substring(start, end+1); // use end+1 for substring, because end index is not included
    }

    private int expand(String s, int l, int r) {
        while(l>=0 && r<s.length() && s.charAt(l)==s.charAt(r)) {
            l--;
            r++;
        }
        return r-l-1;  // this is -1 for "accb", "xacay"
    }

    public static void main(String[] args) {
        Q005 q = new Q005();
        System.out.println(q.longestPalindrome("abba"));
    }
}

File: Algos/Q005.py
# 这个题按照我的做法，并没有performance的问题，我觉得很直观
# 如果需要改进，那就返回两个index及len，而不是substring
class Solution:  # N^2，没有更快时间
    def longestPalindrome(self, s: str) -> str:
        if s is None or len(s) == 0:
            return ""
        maxRes = ""
        for i in range(len(s)):
            s1 = self.expand(s, i, i)
            s2 = self.expand(s, i, i+1)
            res = s1 if len(s1) > len(s2) else s2
            maxRes = res if len(res) > len(maxRes) else maxRes
        return maxRes

    def expand(self, s, left, right):
        resLeft = -1
        resRight = -1
        while left>=0 and right<len(s) and s[left] == s[right]:
            resLeft = left
            resRight = right
            left -= 1
            right += 1
        if resLeft != -1 and resRight != -1:
            return s[resLeft: resRight+1]
        return ""


# 这个改进版本使用index, 时间没有明显区别
class Solution2:
    def longestPalindrome(self, s: str) -> str:
        if s is None or len(s) == 0:
            return ""
        maxLen = 0
        left = 0
        right = 0
        for i in range(len(s)):
            len1, l1, r1 = self.expand(s, i, i)
            len2, l2, r2 = self.expand(s, i, i+1)
            if len1 > maxLen:
                maxLen = len1
                left = l1
                right = r1
            if len2 > maxLen:
                maxLen = len2
                left = l2
                right = r2
        return s[left: right+1]

    def expand(self, s, left, right):
        resLeft = -1
        resRight = -1
        while left>=0 and right<len(s) and s[left] == s[right]:
            resLeft = left
            resRight = right
            left -= 1
            right += 1
        if resLeft == -1 and resRight == -1:
            return 0, resLeft, resRight
        else:
            return resRight-resLeft+1, resLeft, resRight


print(Solution2().longestPalindrome("babad"))
print(Solution2().longestPalindrome("cbbd"))


